packages/eslint-plugin-stories/src/rules/csf-v3-types.ts

```diff
@@ -0,0 +1,138 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/experimental-utils';
+import isStories from '../utils/isStories';
+
+/**
+ * Enforce specifying an explicit type for story objects where necessary for `composeStories` and
+ * Storybook to correctly infer their args.
+ *
+ * For example, instead of
+ *
+ *   const Primary = {
+ *     render: (args) => { ... },
+ *   }
+ *
+ * this rules enforces adding a type annotation
+ *
+ *   const Primary: StoryObj<Args> = {
+ *     render: (args) => { ... },
+ *   }
+ *
+ * Only detects instances where we're sure a type annotation is needed. For example, when there's a
+ * render function defined that takes args (as in the examples above).
+ */
+const rule: TSESLint.RuleModule<'specifyType'> = {
+  meta: {
+    messages: {
+      specifyType:
+        'Specify a type of `StoryObj<Args>` for object stories, so that `composeStories` and Storybook can infer the correct args',
+    },
+    type: 'suggestion',
+    schema: [],
+  },
+  create(context) {
+    if (!isStories(context.getFilename())) {
+      return {};
+    }
+
+    return {
+      // Examine any named exports. In a stories file, these are likely to be stories.
+      // i.e. `export const foo = {}`
+      ExportNamedDeclaration(node) {
+        const story = StoryObject.parse(node);
```

Comment:
Inspried by https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/ and https://www.oreilly.com/library/view/code-that-fits/9780137464302/